File: webApp/src/api/enums.ts
/**
 * Wire-enum ordinal to app-string mappers.
 *
 * The generated OpenAPI schema encodes every Go enum as its integer ordinal (iota
 * order). The SPA view/state-machine logic is written against stable lowerCamel
 * strings, so the api layer converts at the wire boundary (hooks/wire.ts). Ordinals
 * are the authoritative Go iota order (server internal manager contract.gen.go and
 * resourceaccess projectstate). Every integer enum carries an unknown zero value.
 */
import type {
  ArtifactKind,
  ArtifactKindFull,
  CiStatus,
  ProjectArtifactKind,
  ProjectPhase,
  ProjectSessionStage,
  ReviewDecision,
  SDPDecision,
  SessionStage,
} from './types';
import type { ConstructionStage, PipelinePhase, OverrideKind } from './types';
import type { RuntimePhase } from './operationsTypes';
import type { components } from './schema';

type Schemas = components['schemas'];

// --- ArtifactKind (Phase-1 + full, shared 0..16 ordering) ------------------

const ARTIFACT_KIND_BY_ORDINAL: readonly ArtifactKindFull[] = [
  'mission',
  'glossary',
  'scrubbedRequirements',
  'volatilities',
  'coreUseCases',
  'system',
  'operationalConcepts',
  'standardCheck',
  'planningAssumptions',
  'activityList',
  'network',
  'normalSolution',
  'subcriticalSolution',
  'compressedSolution',
  'decompressedSolution',
  'riskModel',
  'sdpReview',
];

const ARTIFACT_KIND_TO_ORDINAL: Record<ArtifactKindFull, number> =
  ARTIFACT_KIND_BY_ORDINAL.reduce<Record<string, number>>((acc, kind, i) => {
    acc[kind] = i;
    return acc;
  }, {});

export function artifactKindFromOrdinal(ordinal: number): ArtifactKindFull {
  return ARTIFACT_KIND_BY_ORDINAL[ordinal] ?? 'mission';
}

export function artifactKindToOrdinal(
  kind: ArtifactKindFull
): Schemas['SystemDesignArtifactKind'] {
  return ARTIFACT_KIND_TO_ORDINAL[kind] as Schemas['SystemDesignArtifactKind'];
}

/** Phase-1 narrowing — the same table, typed back to the Phase-1 union. */
export function systemArtifactKindFromOrdinal(ordinal: number): ArtifactKind {
  return artifactKindFromOrdinal(ordinal) as ArtifactKind;
}

/** Phase-2 narrowing — the same table, typed back to the Phase-2 union. */
export function projectArtifactKindFromOrdinal(ordinal: number): ProjectArtifactKind {
  return artifactKindFromOrdinal(ordinal) as ProjectArtifactKind;
}

// --- System-design SessionStage (0..7) -------------------------------------

const SESSION_STAGE_BY_ORDINAL: readonly SessionStage[] = [
  'unknown',
  'drafting',
  'awaitingReview',
  'redrafting',
  'committed',
  'withdrawn',
  'refused',
  'draftFailed',
];

export function sessionStageFromOrdinal(ordinal: number): SessionStage {
  return SESSION_STAGE_BY_ORDINAL[ordinal] ?? 'unknown';
}

// --- Project-design SessionStage (0..8) ------------------------------------

const PROJECT_SESSION_STAGE_BY_ORDINAL: readonly ProjectSessionStage[] = [
  'unknown',
  'drafting',
  'assemblingSdp',
  'awaitingReview',
  'redrafting',
  'committed',
  'withdrawn',
  'refused',
  'draftFailed',
];

export function projectSessionStageFromOrdinal(ordinal: number): ProjectSessionStage {
  return PROJECT_SESSION_STAGE_BY_ORDINAL[ordinal] ?? 'unknown';
}

// --- ReviewDecision / SDPDecision (request bodies) -------------------------

const REVIEW_DECISION_TO_ORDINAL: Record<ReviewDecision, number> = {
  approve: 1,
  reject: 2,
  withdraw: 3,
};

export function reviewDecisionToOrdinal(
  decision: ReviewDecision
): Schemas['SystemDesignReviewDecision'] {
  return REVIEW_DECISION_TO_ORDINAL[decision] as Schemas['SystemDesignReviewDecision'];
}

const SDP_DECISION_TO_ORDINAL: Record<SDPDecision, number> = {
  commit: 1,
  rejectAll: 2,
};

export function sdpDecisionToOrdinal(decision: SDPDecision): Schemas['ProjectDesignSDPDecision'] {
  return SDP_DECISION_TO_ORDINAL[decision] as Schemas['ProjectDesignSDPDecision'];
}

// --- ProjectPhase (0..2) ---------------------------------------------------

const PROJECT_PHASE_BY_ORDINAL: readonly ProjectPhase[] = [
  'systemDesign',
  'projectDesign',
  'construction',
];

export function projectPhaseFromOrdinal(ordinal: number): ProjectPhase {
  return PROJECT_PHASE_BY_ORDINAL[ordinal] ?? 'unknown';
}

// --- Construction enums ----------------------------------------------------

const CONSTRUCTION_STAGE_BY_ORDINAL: readonly ConstructionStage[] = [
  'unknown',
  'dispatching',
  'pipelineRunning',
  'reviewing',
  'awaitingTakeover',
  'paused',
  'exited',
];

export function constructionStageFromOrdinal(ordinal: number): ConstructionStage {
  return CONSTRUCTION_STAGE_BY_ORDINAL[ordinal] ?? 'unknown';
}

const PIPELINE_PHASE_BY_ORDINAL: readonly PipelinePhase[] = [
  'unknown',
  'pending',
  'running',
  'succeeded',
  'failed',
  'failed', // 5 = cancelled → surfaced as failed (the app has no distinct cancelled state)
];

export function pipelinePhaseFromOrdinal(ordinal: number): PipelinePhase {
  return PIPELINE_PHASE_BY_ORDINAL[ordinal] ?? 'unknown';
}

const OVERRIDE_KIND_TO_ORDINAL: Record<OverrideKind, number> = {
  takeover: 1,
  retry: 2,
  skip: 3,
  reassign: 4,
};

export function overrideKindToOrdinal(kind: OverrideKind): Schemas['ConstructionOverrideKind'] {
  return OVERRIDE_KIND_TO_ORDINAL[kind] as Schemas['ConstructionOverrideKind'];
}

// --- Project head-state row enums ------------------------------------------

/** ProjectCICheckState (0 Pending, 1 Success, 2 Failure) → app CiStatus. */
export function ciStatusFromOrdinal(ordinal: number): CiStatus {
  switch (ordinal) {
    case 1:
      return 'success';
    case 2:
      return 'failed';
    default:
      return 'in_progress';
  }
}

/** ProjectActivityType (0 service,1 frontend,2 testing,3 deployment,4 documentation). */
export function activityRowKindFromOrdinal(ordinal: number): 'service' | 'frontend' | 'testing' {
  switch (ordinal) {
    case 1:
      return 'frontend';
    case 2:
      return 'testing';
    default:
      return 'service';
  }
}

/** ProjectActivityBuildStatus (0 in-construction,1 in-review,2 integrated,3 failed). */
export function buildStatusRowFromOrdinal(
  ordinal: number
): 'integrated' | 'in-review' | 'in-construction' {
  switch (ordinal) {
    case 1:
      return 'in-review';
    case 2:
      return 'integrated';
    default:
      // 0 in-construction; 3 failed surfaces as in-construction (no terminal-fail row state).
      return 'in-construction';
  }
}

// --- Operations enums ------------------------------------------------------

/** OperationsRuntimeStatusSeam (0 unknown,1 pending,2 healthy,3 degraded,4 withdrawn). */
export function runtimePhaseFromOrdinal(ordinal: number): RuntimePhase {
  switch (ordinal) {
    case 1:
      return 'Pending';
    case 2:
      return 'Running';
    case 3:
      return 'Degraded';
    case 4:
      return 'Withdrawn';
    default:
      return 'Unknown';
  }
}

/** OperationsAutoscalerMode (0 unknown,1 auto,2 manual). */
export function autoscalerModeFromOrdinal(ordinal: number): string {
  switch (ordinal) {
    case 1:
      return 'Auto';
    case 2:
      return 'Manual';
    default:
      return 'Unknown';
  }
}

/** OperationsAutoscaleAction (0 noChange,1 scaleUp,2 scaleDown,3 pause,4 resume). */
export function autoscaleActionFromOrdinal(ordinal: number): string {
  switch (ordinal) {
    case 1:
      return 'scaleUp';
    case 2:
      return 'scaleDown';
    case 3:
      return 'pause';
    case 4:
      return 'resume';
    default:
      return 'noChange';
  }
}

// --- Operations desired-state-change ordinals (request bodies) -------------

/** OperationsDesiredStateReason ordinals. */
export const REASON_DEPLOY_AFTER_CONSTRUCTION = 1;
export const REASON_OPERATOR = 2;

/** OperationsPatchKind ordinals. */
export const PATCH_FULL_BUNDLE = 1;
export const PATCH_SCALE = 2;
export const PATCH_POLICY = 3;

File: systemtests/internal/harness/enums.go
package harness

// enums.go is the SINGLE source that bridges the harness's transport-agnostic
// STRING vocabulary (the wire enum names the use-case steps assert on:
// "mission", "drafting", "assemblingSdp", ...) to the generated SDK's TYPED enum
// consts (internal/sdk). The SDK owns the wire ordinals now — every table below
// keys on an sdk const, so there is NO hand-maintained ordinal literal anywhere
// in the harness (the 11 ordinal tables httptransport.go used to carry are
// deleted). Both wire transports (httptransport.go, mcptransport.go) delegate
// their enum encode/decode here, so the two surfaces speak identical strings.
//
// Wire names are the RAW wire vocabulary (e.g. PipelineCancelled stays
// "cancelled", RuntimeStatusHealthy stays "healthy") — the harness asserts on
// the literal contract, unlike webApp which folds/renames some for display.
// Only ProjectSessionStage ordinal 2 is hand-mapped: the SDK varname
// ProjectDesignStageAssemblingSDP would mechanically lower-first to
// "assemblingSDP", but the wire contract is "assemblingSdp".

import (
	"github.com/mixofreality-studio/archistrator/systemtests/internal/sdk"
)

// TestOwner is the fixed OwnerScope the harness mints every project under. Owner
// is a required, non-empty CreateProject arg but is NOT consulted by
// authenticatedOnlyPDP (any authenticated principal may act on any resource —
// see server/cmd/server/authz.go) and the Transport interface's CreateProject
// has no owner parameter, so a single constant value is sufficient for a
// black-box test. Exported so a test can pass the SAME scope to ListProjects.
const TestOwner = "systemtest"

// testOwner is the unexported alias every existing call site in this package uses.
const testOwner = TestOwner

// invert flips a const→wire map into a wire→const map for the encode direction.
func invert[K comparable, V comparable](m map[K]V) map[V]K {
	out := make(map[V]K, len(m))
	for k, v := range m {
		out[v] = k
	}
	return out
}

// --- ArtifactKind (shared by system-design + project-design) -----------------

var artifactKindWire = map[sdk.ArtifactKind]string{
	sdk.KindMission:              "mission",
	sdk.KindGlossary:             "glossary",
	sdk.KindScrubbedRequirements: "scrubbedRequirements",
	sdk.KindVolatilities:         "volatilities",
	sdk.KindCoreUseCases:         "coreUseCases",
	sdk.KindSystem:               "system",
	sdk.KindOperationalConcepts:  "operationalConcepts",
	sdk.KindStandardCheck:        "standardCheck",
	sdk.KindPlanningAssumptions:  "planningAssumptions",
	sdk.KindActivityList:         "activityList",
	sdk.KindNetwork:              "network",
	sdk.KindNormalSolution:       "normalSolution",
	sdk.KindSubcriticalSolution:  "subcriticalSolution",
	sdk.KindCompressedSolution:   "compressedSolution",
	sdk.KindDecompressedSolution: "decompressedSolution",
	sdk.KindRiskModel:            "riskModel",
	sdk.KindSdpReview:            "sdpReview",
}

var artifactKindByWire = invert(artifactKindWire)

// artifactKind encodes a wire kind name into the SDK enum a request expects. An
// unknown name maps to KindMission (0) — callers only pass known names.
func artifactKind(name string) sdk.ArtifactKind { return artifactKindByWire[name] }

// artifactKindNameOf decodes an SDK ArtifactKind into its wire name, defaulting
// to "mission" for an out-of-range value (matches the retired ordinal table).
func artifactKindNameOf(k sdk.ArtifactKind) string {
	if name, ok := artifactKindWire[k]; ok {
		return name
	}
	return "mission"
}

// ArtifactKindName decodes a plan step's numeric ArtifactKind ordinal (0..16,
// shared by systemDesignManager and projectDesignManager) into the wire kind
// name a Transport method's `kind string` parameter expects.
func ArtifactKindName(ordinal int) string { return artifactKindNameOf(sdk.ArtifactKind(ordinal)) }

// --- ReviewDecision ----------------------------------------------------------

var reviewDecisionWire = map[sdk.ReviewDecision]string{
	sdk.ReviewApprove:  "approve",
	sdk.ReviewReject:   "reject",
	sdk.ReviewWithdraw: "withdraw",
}

var reviewDecisionByWire = invert(reviewDecisionWire)

func reviewDecision(name string) sdk.ReviewDecision { return reviewDecisionByWire[name] }

// ReviewDecisionName decodes a plan step's numeric ReviewDecision ordinal (0
// unknown,1 approve,2 reject,3 withdraw) into the decision name
// Transport.SubmitReview / SubmitProjectReview expect.
func ReviewDecisionName(ordinal int) string { return reviewDecisionWire[sdk.ReviewDecision(ordinal)] }

// --- SDPDecision -------------------------------------------------------------

var sdpDecisionWire = map[sdk.SDPDecision]string{
	sdk.SDPCommit:    "commit",
	sdk.SDPRejectAll: "rejectAll",
}

var sdpDecisionByWire = invert(sdpDecisionWire)

func sdpDecision(name string) sdk.SDPDecision { return sdpDecisionByWire[name] }

// SDPDecisionName decodes a plan step's numeric SDPDecision ordinal (0
// unknown,1 commit,2 rejectAll) into the decision name
// Transport.SubmitSDPDecision expects.
func SDPDecisionName(ordinal int) string { return sdpDecisionWire[sdk.SDPDecision(ordinal)] }

// --- PhaseDecision (construction) --------------------------------------------

var phaseDecisionWire = map[sdk.PhaseDecision]string{
	sdk.PhaseApprove:  "approve",
	sdk.PhaseSendBack: "sendBack",
}

var phaseDecisionByWire = invert(phaseDecisionWire)

func phaseDecision(name string) sdk.PhaseDecision { return phaseDecisionByWire[name] }

// --- DesiredStateReason / PatchKind (operations) -----------------------------

var desiredStateReasonWire = map[sdk.DesiredStateReason]string{
	sdk.ReasonDeployAfterConstruction: "deployAfterConstruction",
	sdk.ReasonOperator:                "operator",
	sdk.ReasonAutoscale:               "autoscale",
	sdk.ReasonDelinquency:             "delinquency",
}

var desiredStateReasonByWire = invert(desiredStateReasonWire)

func desiredStateReason(name string) sdk.DesiredStateReason { return desiredStateReasonByWire[name] }

var patchKindWire = map[sdk.PatchKind]string{
	sdk.PatchFullBundle: "fullBundle",
	sdk.PatchScale:      "scale",
	sdk.PatchPolicy:     "policy",
}

var patchKindByWire = invert(patchKindWire)

func patchKind(name string) sdk.PatchKind { return patchKindByWire[name] }

// --- stage decoders (response-only) ------------------------------------------

var systemStageWire = map[sdk.SessionStage]string{
	sdk.SessionStageUnknown: "unknown",
	sdk.StageDrafting:       "drafting",
	sdk.StageAwaitingReview: "awaitingReview",
	sdk.StageRedrafting:     "redrafting",
	sdk.StageCommitted:      "committed",
	sdk.StageWithdrawn:      "withdrawn",
	sdk.StageRefused:        "refused",
	sdk.StageDraftFailed:    "draftFailed",
}

func systemStageName(s sdk.SessionStage) string {
	if name, ok := systemStageWire[s]; ok {
		return name
	}
	return "unknown"
}

// projectStageWire has ONE more stage than the system-design enum
// (assemblingSdp at ordinal 2). ProjectDesignStageAssemblingSDP is HAND-MAPPED
// to "assemblingSdp" — the mechanical lower-first of the SDK varname would give
// "assemblingSDP" (mirrors webApp/scripts/gen-enums.mjs NON_MECHANICAL).
var projectStageWire = map[sdk.ProjectSessionStage]string{
	sdk.ProjectSessionStageUnknown: "unknown",
	sdk.ProjectStageDrafting:       "drafting",
	sdk.ProjectStageAssemblingSDP:  "assemblingSdp",
	sdk.ProjectStageAwaitingReview: "awaitingReview",
	sdk.ProjectStageRedrafting:     "redrafting",
	sdk.ProjectStageCommitted:      "committed",
	sdk.ProjectStageWithdrawn:      "withdrawn",
	sdk.ProjectStageRefused:        "refused",
	sdk.ProjectStageDraftFailed:    "draftFailed",
}

func projectStageName(s sdk.ProjectSessionStage) string {
	if name, ok := projectStageWire[s]; ok {
		return name
	}
	return "unknown"
}

var constructionStageWire = map[sdk.ConstructionStage]string{
	sdk.ConstructionStageUnknown: "unknown",
	sdk.StageDispatching:         "dispatching",
	sdk.StagePipelineRunning:     "pipelineRunning",
	sdk.StageReviewing:           "reviewing",
	sdk.StageAwaitingTakeover:    "awaitingTakeover",
	sdk.StagePaused:              "paused",
	sdk.StageExited:              "exited",
	sdk.StageAwaitingApproval:    "awaitingApproval",
}

func constructionStageName(s sdk.ConstructionStage) string {
	if name, ok := constructionStageWire[s]; ok {
		return name
	}
	return "unknown"
}

var pipelinePhaseWire = map[sdk.PipelinePhase]string{
	sdk.PipelinePhaseUnknown: "unknown",
	sdk.PipelinePending:      "pending",
	sdk.PipelineRunning:      "running",
	sdk.PipelineSucceeded:    "succeeded",
	sdk.PipelineFailed:       "failed",
	sdk.PipelineCancelled:    "cancelled",
}

func pipelinePhaseName(p sdk.PipelinePhase) string {
	if name, ok := pipelinePhaseWire[p]; ok {
		return name
	}
	return "unknown"
}

var runtimeStatusWire = map[sdk.RuntimeStatusSeam]string{
	sdk.RuntimeStatusUnknown:   "unknown",
	sdk.RuntimeStatusPending:   "pending",
	sdk.RuntimeStatusHealthy:   "healthy",
	sdk.RuntimeStatusDegraded:  "degraded",
	sdk.RuntimeStatusWithdrawn: "withdrawn",
}

func runtimeStatusName(r sdk.RuntimeStatusSeam) string {
	if name, ok := runtimeStatusWire[r]; ok {
		return name
	}
	return "unknown"
}

// decodeMissingArtifacts decodes a PhaseAdvanceResult's []ArtifactKind into the
// wire kind names a caller asserts on.
func decodeMissingArtifacts(kinds []sdk.ArtifactKind) []string {
	out := make([]string, 0, len(kinds))
	for _, k := range kinds {
		out = append(out, artifactKindNameOf(k))
	}
	return out
}

// strPtrVal dereferences an optional wire string pointer, defaulting to "".
func strPtrVal(s *string) string {
	if s == nil {
		return ""
	}
	return *s
}

// ---------------------------------------------------------------------------
// STAGE 4a — THE ACTIVITY/TASK ADDRESS OF AN ARTIFACT KIND
//
// The twelve-op deliveryManager addresses work by {activityID, taskID}, not by
// artifact kind: a draft is a DISPATCH task of a design activity and a verdict is
// its REVIEW task. The tables below are the harness's copy of that address, and
// they mirror the committed method-assets lifecycles exactly — requirements owns
// the four Phase-1 drafts, architecture owns the System draft, and projectDesign
// owns the single computed sdpReview gate.
// ---------------------------------------------------------------------------

// designActivityFor is the activity that owns an artifact kind's tasks.
func designActivityFor(kind string) sdk.ActivityID {
	switch kind {
	case "mission", "glossary", "scrubbedRequirements", "volatilities", "coreUseCases":
		return "requirements"
	case "system", "operationalConcepts", "standardCheck":
		return "architecture"
	default:
		return "projectDesign"
	}
}

// draftTaskFor is the DISPATCH task that produces an artifact kind.
func draftTaskFor(kind string) string {
	if t, ok := designDraftTasks[kind]; ok {
		return t
	}
	return "sdpReview"
}

// reviewTaskFor is the REVIEW task that judges an artifact kind's draft.
func reviewTaskFor(kind string) string {
	if t, ok := designReviewTasks[kind]; ok {
		return t
	}
	return "sdpReview"
}

var designDraftTasks = map[string]string{
	"mission":      "missionDraft",
	"glossary":     "glossaryDraft",
	"volatilities": "volatilitiesDraft",
	"coreUseCases": "coreUseCasesDraft",
	"system":       "architectureDraft",
}

var designReviewTasks = map[string]string{
	"mission":      "missionReview",
	"glossary":     "glossaryReview",
	"volatilities": "volatilitiesReview",
	"coreUseCases": "coreUseCasesReview",
	"system":       "architectureReview",
}

// sdpReviewDecision maps the SDP vocabulary onto the merged ReviewDecision: the M0
// gate's commit is an approve and its reject-all is a reject.
func sdpReviewDecision(name string) sdk.ReviewDecision {
	if name == "rejectAll" {
		return sdk.ReviewReject
	}
	return sdk.ReviewApprove
}

// phaseReviewDecision maps the construction gate's vocabulary onto the merged
// ReviewDecision: approve stays approve and sendBack is a reject.
func phaseReviewDecision(name string) sdk.ReviewDecision {
	if name == "sendBack" {
		return sdk.ReviewReject
	}
	return sdk.ReviewApprove
}

// activityViewStateName is the ActivityView's coarse state as its raw wire string —
// the enum is a STRING enum (notStarted | running | awaitingReview | done | failed), so
// the value IS the name and no ordinal table is needed.
func activityViewStateName(s sdk.ActivityViewState) string { return string(s) }

// ---------------------------------------------------------------------------
// STAGE 4a — the reverse of designActivityFor/draftTaskFor: the plan's steps
// address a TASK, and the harness's transport-agnostic vocabulary still speaks
// artifact kinds, so the runner needs the task -> kind direction too.
// ---------------------------------------------------------------------------

// ArtifactKindForTask is the artifact a design lifecycle task is about — "" when the
// task id names no design artifact (a construction lifecycle phase, which the
// construction gate takes verbatim).
func ArtifactKindForTask(taskID string) string {
	for kind, draft := range designDraftTasks {
		if draft == taskID {
			return kind
		}
	}
	for kind, review := range designReviewTasks {
		if review == taskID {
			return kind
		}
	}
	if taskID == "sdpReview" {
		return "sdpReview"
	}
	return ""
}

// IsPhase2ArtifactKind reports whether a kind belongs to Project Design (Phase 2) — the
// split QueryProjectView's session arm routes on, and the one the harness's two
// GetSessionState reads mirror.
func IsPhase2ArtifactKind(kind string) bool {
	return int(artifactKind(kind)) >= int(sdk.KindPlanningAssumptions)
}

// ReviewAdvanceOrdinal is the merged ReviewDecision's phase-seal ordinal, exported so
// the generated-table runner can recognise an advance without re-declaring the enum.
const ReviewAdvanceOrdinal = sdk.ReviewAdvance
